### PR TITLE
Domain audit file

### PR DESCRIPTION
**Pull request by sgrigson**
Allows for the creation of a Blocklist Audit File.

This audit file can be used to return metadata about each domain match, such as the number of matches by count or percentage.

(This was primarily done for my own needs, but it should be easy enough to ignore for users who do not want to use this feature, as by default it's not enabled.)

### CODE REVIEW

File: src/fediblockhole/__init__.py [hunk 1]
```diff
@@ -701,6 +745,9 @@ def augment_args(args, tomldata: str=None):
     if not args.savedir:
         args.savedir = conf.get('savedir', '/tmp')
 
+    if not args.blocklist_auditfile:
+        args.blocklist_auditfile = conf.get('blocklist_auditfile', '')
```
Comment (jpwarren) [inline src/fediblockhole/__init__.py:749]: I think this should return None if no `blocklist_auditfile` is set.

File: src/fediblockhole/__init__.py [hunk 2]
```diff
@@ -183,7 +183,8 @@ def fetch_from_instances(sources: dict,
 
 def merge_blocklists(blocklists: list[Blocklist], mergeplan: str='max',
     threshold: int=0,
-    threshold_type: str='count') -> Blocklist:
+    threshold_type: str='count',
+    save_block_audit_file: str='') -> Blocklist:
```
Comment (jpwarren) [inline src/fediblockhole/__init__.py:187]: Why does this string default to empty? Wouldn't a `None` be a better indicator that no audit file is needed?
I'd be worried I'd accidentally clobber some other file in the working directory by opening an empty string for writing or something.

File: src/fediblockhole/__init__.py [hunk 3]
```diff
@@ -238,6 +242,18 @@ def merge_blocklists(blocklists: list[Blocklist], mergeplan: str='max',
                 block = apply_mergeplan(block, newblock, mergeplan)
             merged.blocks[block.domain] = block
 
+        if len(save_block_audit_file) > 0:
```
Comment (jpwarren) [inline src/fediblockhole/__init__.py:245]: Yeah, I think this check should be `if save_block_audit_file:` which will match on None or empty list/string.

Could more explicitly require `if save_block_audit_file is not None:` but a wider capture of non-useful settings (like empty strings) here probably guards more against accidentally opening a wrong logfile, and matching the user intent.

I think I'd find the first method easier to interpret after a while away from the code. It took me a moment to figure out why this needed to be a positive-length item, and it's more that it needs to be something that can be opened as a file. Let's be kind to future us.

**Review by jpwarren — CHANGES_REQUESTED** [review]
Good addition, I like it.

Small change to the semantics requested: make the default audit filepath `None` instead of an empty string, then check for existence/truthiness to see if you need to save audit info.

**Comment by sgrigson** [comment]
Good points! Updated and ready for another review.

I'm very new to Python (mostly working with Golang lately), so appreciate the best practice tip there about 'None'.

**Review by jpwarren — COMMENTED** [review]
Looks good!

**Pull request merged**